packages/flutter_test/lib/src/tree_traversal.dart

```diff
@@ -0,0 +1,167 @@
+/// one, for example the results of calling `where` on this iterable
+/// are also cached.
+Iterable<Element> collectAllElementsFrom(
+  Element rootElement, {
+  // TODO(pdblasi-google): Deprecate `skipOffstage`. @Deprecated(
+  //   'Use includeOffstage instead. '
+  //   'This allows us to be consistent with the new FinderBase API. '
+  //   'This feature was deprecated after v3.11.0-0.0.pre.'
+  // )
+  bool? skipOffstage,
+  bool includeOffstage = false,
+}) {
+  if (skipOffstage != null) {
+    includeOffstage = !skipOffstage;
+  }
+  return CachingIterable<Element>(_DepthFirstElementTreeIterator(rootElement, includeOffstage));
+}
+
+/// Provides an iterable that efficiently returns all the [SemanticsNode]s
+/// rooted at the given [SemanticsNode]. See [CachingIterable] for details.
+///
+/// This method must be called again if the semantics change. You cannot call
+/// this function once, then reuse the iterable after having changed the state
+/// of the tree, because the iterable returned by this function caches the
+/// results and only walks the tree once.
+///
+/// The same applies to any iterable obtained indirectly through this
+/// one, for example the results of calling `where` on this iterable
+/// are also cached.
+Iterable<SemanticsNode> collectAllSemanticsNodesFrom(
+  SemanticsNode root, {
+    DebugSemanticsDumpOrder order = DebugSemanticsDumpOrder.traversalOrder,
+    bool includeRoot = true,
+  }) {
+    return CachingIterable<SemanticsNode>(_DepthFirstSemanticsTreeIterator(root, order, includeRoot: includeRoot));
+}
+
+/// Provides a recursive, efficient, depth first search of a tree.
+///
+/// This iterator executes a depth first search as an iterable, and iterates in
+/// a left to right order:
+///
+///       1
+///     /   \
+///    2     3
+///   / \   / \
+///  4   5 6   7
+///
+/// Will iterate in order (1), 2, 4, 5, 3, 6, 7. The root element (1) can be
+/// included by passing `includeRoot` as `true` during construction.
+abstract class _DepthFirstTreeIterator<TItem> implements Iterator<TItem> {
+  _DepthFirstTreeIterator(TItem root, {bool includeRoot = true}) {
+    if (includeRoot) {
+      _stack.add(root);
+    } else {
+      _fillStack(_collectChildren(root));
+    }
+  }
+
+  @override
+  TItem get current => _current!;
+  late TItem _current;
+
+  final List<TItem> _stack = <TItem>[];
+
+  @override
+  bool moveNext() {
+    if (_stack.isEmpty) {
+      return false;
+    }
+
+    _current = _stack.removeLast();
+    _fillStack(_collectChildren(_current));
+    return true;
+  }
+
+  /// Fills the stack in such a way that the next element of a depth first
+  /// traversal is easily and efficiently accessible when calling `moveNext`.
+  void _fillStack(List<TItem> children) {
+    // We reverse the list of children so we don't have to do use expensive
+    // `insert` or `remove` operations, and so the order of the traversal
+    // is depth first when built lazily through the iterator.
+
```

Comment:
pedantic nit: I'd add `//` in this line since this is conceptually a single comment but divided into 2 paragraphs.

Reply:
Done!